tests/specs/infinite-gas.k

```diff
@@ -1,35 +1,96 @@
+    rule I1         <=Int I2 +Int I3 => I1 -Int I3 <=Int I2         requires #isConcrete(I1) andBool #isConcrete(I3) [simplification]
+
+    rule ((I1 +Int I2) +Int I3) +Int I4 => (I1 +Int I3) +Int (I2 +Int I4) requires #isConcrete(I2) andBool #isConcrete(I4) andBool notBool (#isConcrete(I1) orBool #isConcrete(I3)) [simplification]
+    rule ((I1 +Int I2) -Int I3) +Int I4 => (I1 -Int I3) +Int (I2 +Int I4) requires #isConcrete(I2) andBool #isConcrete(I4) andBool notBool (#isConcrete(I1) orBool #isConcrete(I3)) [simplification]
+    rule (I1 +Int I2) -Int (I3 +Int I4) => (I1 -Int I3) +Int (I2 -Int I4) requires #isConcrete(I2) andBool #isConcrete(I4) andBool notBool (#isConcrete(I1) orBool #isConcrete(I3)) [simplification]
+endmodule
+
+module INFINITE-GAS-COMMON
     imports EVM
 
+ // Generic rules
+
+    rule I -Int I => 0                                              [simplification]
+    rule I1 -Int (I1 -Int I2) => I2                                 [simplification]
+    rule (notBool (A andBool B)) andBool A => (notBool B) andBool A [simplification]
 
  // Symbolic Gas
 
     syntax Int ::= #gas ( Int , Int , Int )  [function]  // startGas, nonMemory, memory
  //-----------------------------------------------------------------------------------
 
- //
- // Infinite Gas
- // Here we use the construct `#gas` to count _up_ gas used instead of counting down.
- // This allows (i) computing final gas used, and (ii) never stopping because of out-of-gas.
- //
-
     syntax Int ::= #gas ( Int ) [function, functional, no-evaluators, klabel(infGas), symbol, smtlib(infGas)]
  // ---------------------------------------------------------------------------------------------------------
     rule #gas(_, _, _) => #gas(0)
 
-    rule #gas(G) -Int G' => #gas(G +Int G') [simplification]
+    rule #gas(G) +Int G' => #gas(G +Int G') requires 0 <Int G' orBool 0 -Int G' <Int #gas(G)  [simplification]
+    rule G +Int #gas(G') => #gas(G +Int G') requires 0 <Int G  orBool 0 -Int G  <Int #gas(G') [simplification]
+
+    rule #gas(G) -Int G' => #gas(G -Int G') requires 0 <Int G' andBool G' <Int #gas(G) [simplification]
+
+    rule #gas(G) *Int G' => #gas(G *Int G') requires 0 <Int G' [simplification]
+    rule G *Int #gas(G') => #gas(G *Int G') requires 0 <Int G  [simplification]
+
+    rule #gas(G) /Int G' => #gas(G /Int G') requires 0 <Int G' andBool G' <Int #gas(G)  [simplification]
+    rule G /Int #gas(G') => 0               requires 0 <Int G  andBool G  <Int #gas(G') [simplification]
+
+    rule #gas(#gas(G)) => #gas(G) [simplification]
+
+    rule #gas(G) -Int #gas(G') => #gas(G -Int G') [simplification]
+
+    rule minInt(#gas(G), #gas(G'))              => #gas(minInt(G, G'))              [simplification]
+    rule #if B #then #gas(G) #else #gas(G') #fi => #gas(#if B #then G #else G' #fi) [simplification]
+
+    rule #gas(_)  <Int #gas(_) => false [simplification]
+    rule #gas(_) <=Int #gas(_) => true  [simplification]
```

Comment:
Can you explain these 2 lemmas in a comment?

Reply:
Also, in what kind of scenarios are they useful?

Reply:
Thanks! All makes sense now. I think it's useful to add your explanation as comment in k file.

Reply:
I added a comment with this explanation to the documentation at the top of the file.